File: Polars.FSharp/GroupByBuilder.fs
namespace Polars.FSharp

open System
open Polars.NET.Core

/// <summary>
/// Intermediate builder for LazyGroupBy operations.
/// Holds the LazyFrame handle (ownership transferred to this builder) and grouping keys.
/// </summary>
[<Sealed>]
type LazyGroupBy internal (lfHandle: LazyFrameHandle, keys: Expr[]) =
    
    let mutable disposed = false
    let mutable havingExpr: Expr option = None
    
    let ownedKeyHandles = 
        keys |> Array.map (fun k -> PolarsWrapper.CloneExpr k.Handle)

    member this.Dispose() =
        if not disposed then
            for h in ownedKeyHandles do
                if not (isNull (box h)) && not h.IsInvalid then 
                    h.Dispose()
            
            if not lfHandle.IsClosed then 
                lfHandle.Dispose()
            
            disposed <- true
            
        GC.SuppressFinalize this

    interface IDisposable with
        member this.Dispose() = this.Dispose()

    // --------------------------------------------------------
    // Constructor
    // --------------------------------------------------------
    
    /// <summary>
    /// Filter groups with a list of predicates after aggregation.
    /// Using this method is equivalent to adding the predicates to the aggregation and filtering afterwards.
    /// </summary>
    member this.Having(predicate: Expr) : LazyGroupBy =
        havingExpr <- Some predicate
        this 

    /// <summary>
    /// Apply aggregations to the group.
    /// </summary>
    member this.Agg(aggs: seq<Expr>) : LazyFrame =
        if disposed then raise (ObjectDisposedException "LazyGroupBy")

        let aggArray = Seq.toArray aggs
        let aggHandles = aggArray |> Array.map (fun a -> PolarsWrapper.CloneExpr a.Handle)
        let keysForRust = ownedKeyHandles |> Array.map (fun k -> PolarsWrapper.CloneExpr k)

        let havingHandle = 
            match havingExpr with
            | Some expr -> PolarsWrapper.CloneExpr(expr.Handle)
            | None -> null
            
        let resHandle = PolarsWrapper.LazyGroupByAgg(lfHandle, keysForRust, aggHandles, havingHandle)
        new LazyFrame(resHandle)

    member this.Agg([<ParamArray>] aggs: Expr[]) : LazyFrame =
        this.Agg(aggs :> seq<Expr>)

    // --------------------------------------------------------
    // Syntactic Sugar
    // --------------------------------------------------------

    /// <summary> Count the number of values in each group. </summary>
    member this.Count() = this.Agg [ Expr.All().Count() ]

    /// <summary> Aggregate all columns into lists. </summary>
    member this.All() = this.Agg [ Expr.All() ]

    /// <summary> Aggregate the first values in the group. </summary>
    member this.First(?ignoreNulls: bool) =
        let ig = defaultArg ignoreNulls false
        this.Agg [ Expr.All().First ig ]

    /// <summary> Aggregate the last values in the group. </summary>
    member this.Last(?ignoreNulls: bool) =
        let ig = defaultArg ignoreNulls false
        this.Agg [ Expr.All().Last ig ]

    /// <summary> Get the first n rows of each group. </summary>
    member this.Head(?n: int) =
        let count = defaultArg n 10
        let aggregated = this.Agg [ Expr.All().Head count ]
        
        let keyNames = 
            keys 
            |> Array.choose (fun expr -> 
                match expr.Meta.OutputName() with
                | null | "" -> None
                | name -> Some name
            )
            
        aggregated.Explode(Expr.All().ToSelector().Exclude keyNames )

    /// <summary> Get the last n rows of each group. </summary>
    member this.Tail(?n: int) =
        let count = defaultArg n 10
        let aggregated = this.Agg [ Expr.All().Tail(count) ]
        
        let keyNames = 
            keys 
            |> Array.choose (fun expr -> 
                match expr.Meta.OutputName() with
                | null | "" -> None
                | name -> Some name
            )
            
        aggregated.Explode(Expr.All().ToSelector().Exclude keyNames )

    /// <summary> Return the number of rows in each group. </summary>
    member this.Len(?name: string) =
        let colName = defaultArg name "len"
        this.Agg [ Expr.Len().Alias colName ]

    /// <summary> Reduce the groups to the maximal value. </summary>
    member this.Max() = this.Agg [ Expr.All().Max() ]

    /// <summary> Reduce the groups to the minimal value. </summary>
    member this.Min() = this.Agg [ Expr.All().Min() ]

    /// <summary> Reduce the groups to the median value. </summary>
    member this.Median() = this.Agg [ Expr.All().Median() ]

    /// <summary> Reduce the groups to the mean value. </summary>
    member this.Mean() = this.Agg [ Expr.All().Mean() ]

    /// <summary> Count the unique values per group. </summary>
    member this.NUnique() = this.Agg [ Expr.All().NUnique() ]

    /// <summary> Reduce the groups to the sum. </summary>
    member this.Sum() = this.Agg [ Expr.All().Sum() ]

    /// <summary> Compute the quantile per group. </summary>
    member this.Quantile(quantile: float, ?interpolation: QuantileMethod) =
        let interp = defaultArg interpolation QuantileMethod.Linear
        this.Agg [ Expr.All().Quantile(quantile, interp) ]

[<AutoOpen>]
module LazyGroupByExtensions =

    type LazyFrame with

        /// <summary>
        /// Start a GroupBy operation. Returns a builder to apply syntax sugar aggregations (like .Sum(), .Head()).
        /// </summary>
        member this.GroupBy(keys: seq<Expr>, ?having: Expr) : LazyGroupBy =
            let builder = new LazyGroupBy(this.CloneHandle(), Seq.toArray keys)
            match having with
            | Some h -> builder.Having h 
            | None -> builder

        /// <summary>
        /// Start a GroupBy operation using Selectors/IColumnExpr. 
        /// </summary>
        member this.GroupBy(keys: seq<#IColumnExpr>, ?having: Expr) : LazyGroupBy =
            let kExprs = keys |> Seq.collect (fun x -> x.ToExprs()) |> Seq.toArray
            let builder = new LazyGroupBy(this.CloneHandle(), kExprs)
            match having with
            | Some h -> builder.Having(h)
            | None -> builder

        member this.GroupBy([<ParamArray>] keys: Expr[]) : LazyGroupBy =
            this.GroupBy(keys :> seq<Expr>)

        /// <summary>
        /// Group by keys and apply aggregate expressions, optionally filtering groups.
        /// </summary>
        member this.GroupBy(keys: seq<Expr>, aggs: seq<Expr>, ?having: Expr) : LazyFrame =
            use builder = this.GroupBy(keys, ?having = having)
            builder.Agg aggs

        /// <summary>
        /// Group by keys and apply aggregations (Supports Selectors), optionally filtering groups.
        /// </summary>
        member this.GroupBy(keys: seq<#IColumnExpr>, aggs: seq<#IColumnExpr>, ?having: Expr) : LazyFrame =
            let aExprs = aggs |> Seq.collect (fun x -> x.ToExprs())
            use builder = this.GroupBy(keys, ?having = having)
            builder.Agg aExprs

/// <summary>
/// Intermediate builder for eager GroupBy operations.
/// Under the hood, this routes through the Lazy engine to maximize performance.
/// </summary>
type GroupBy internal (df: DataFrame, keys: Expr[]) =

    let mutable disposed = false
    
    let mutable havingExpr: Expr option = None

    let ownedKeyHandles = 
        keys |> Array.map (fun k -> PolarsWrapper.CloneExpr(k.Handle))

    /// <summary> Filter groups with a predicate after aggregation. </summary>
    member this.Having(predicate: Expr) =
        havingExpr <- Some predicate
        this

    member this.Dispose() =
        if not disposed then
            for h in ownedKeyHandles do
                if not (isNull (box h)) && not h.IsInvalid then 
                    h.Dispose()
        
            disposed <- true
            
        GC.SuppressFinalize(this)

    interface IDisposable with
        member this.Dispose() = this.Dispose()

    member this.Agg(aggs: seq<Expr>) : DataFrame =
        if disposed then raise (ObjectDisposedException "GroupBy Builder")

        let safeKeys = ownedKeyHandles |> Array.map (fun h -> new Expr(PolarsWrapper.CloneExpr h))
        
        let lazyBuilder = df.Lazy().GroupBy(safeKeys)
        
        let finalBuilder = 
            match havingExpr with
            | Some h -> lazyBuilder.Having(h)
            | None -> lazyBuilder
            
        use lazyResult = finalBuilder.Agg aggs
        
        for k in safeKeys do (k :> IDisposable).Dispose()

        lazyResult.Collect()

    member this.Agg([<ParamArray>] aggs: Expr[]) : DataFrame =
        this.Agg(aggs :> seq<Expr>)

    member this.Count() = this.Agg [ Expr.All().Count() ]
    member this.All() = this.Agg [ Expr.All() ]
    
    member this.First(?ignoreNulls: bool) =
        let ig = defaultArg ignoreNulls false
        this.Agg [ Expr.All().First ig ]
        
    member this.Last(?ignoreNulls: bool) =
        let ig = defaultArg ignoreNulls false
        this.Agg [ Expr.All().Last ig ]
        
    member this.Len(?name: string) =
        let colName = defaultArg name "len"
        this.Agg [ Expr.Len().Alias colName ]

    member this.Max() = this.Agg [ Expr.All().Max() ]
    member this.Min() = this.Agg [ Expr.All().Min() ]
    member this.Median() = this.Agg [ Expr.All().Median() ]
    member this.Mean() = this.Agg [ Expr.All().Mean() ]
    member this.NUnique() = this.Agg [ Expr.All().NUnique() ]
    member this.Sum() = this.Agg [ Expr.All().Sum() ]

    member this.Quantile(quantile: float, ?interpolation: QuantileMethod) =
        let interp = defaultArg interpolation QuantileMethod.Linear
        this.Agg [ Expr.All().Quantile(quantile, interp) ]

    member this.Head(?n: int) =
        let count = defaultArg n 10
        let aggregatedDf = this.Agg [ Expr.All().Head count ]
        
        let keyNames = 
            keys |> Array.choose (fun expr -> 
                match expr.Meta.OutputName() with
                | null | "" -> None
                | name -> Some name
            )
        aggregatedDf.Explode(Expr.All().ToSelector().Exclude keyNames)

    member this.Tail(?n: int) =
        let count = defaultArg n 10
        let aggregatedDf = this.Agg [ Expr.All().Tail count ]
        let keyNames = 
            keys |> Array.choose (fun expr -> 
                match expr.Meta.OutputName() with
                | null | "" -> None
                | name -> Some name
            )
        aggregatedDf.Explode(Expr.All().ToSelector().Exclude keyNames)

[<AutoOpen>]
module DataFrameGroupByExtensions =

    type DataFrame with
        
        /// <summary> Start a GroupBy operation. </summary>
        member this.GroupBy(keys: seq<Expr>) =
            new GroupBy(this, Seq.toArray keys)

        /// <summary> Start a GroupBy operation with Selectors. </summary>
        member this.GroupBy(keys: seq<#IColumnExpr>) =
            let kExprs = keys |> Seq.collect (fun x -> x.ToExprs()) |> Seq.toArray
            new GroupBy(this, kExprs)

        member this.GroupBy([<ParamArray>] keys: Expr[]) =
            new GroupBy(this, keys)

        /// <summary> Group by keys and apply aggregate expressions. </summary>
        member this.GroupBy(keys: seq<Expr>, aggs: seq<Expr>, ?having: Expr) : DataFrame =
            let builder = this.GroupBy keys
            match having with
            | Some h -> builder.Having(h).Agg aggs
            | None -> builder.Agg aggs
        
        member this.GroupBy(keys: Expr, aggs: seq<Expr>, ?having: Expr) : DataFrame =
            this.GroupBy([keys], aggs, ?having = having)

        member this.GroupBy(keys: seq<Expr>, aggs: Expr, ?having: Expr) : DataFrame =
            this.GroupBy(keys, [aggs], ?having = having)

        member this.GroupBy(keys: Expr, aggs: Expr, ?having: Expr) : DataFrame =
            this.GroupBy([keys], [aggs], ?having = having)

File: Polars.CSharp/Operations/GroupByBuilder.cs
#pragma warning disable CS1591 
using System.Collections;
using Cs = Polars.CSharp.Polars.Selectors;
using Pl = Polars.CSharp.Polars;

namespace Polars.CSharp;
/// <summary>
/// Unified builder for DataFrame GroupBy operations (Standard, Dynamic, Rolling)
/// </summary>
public class GroupByBuilder:IEnumerable<(object[] Key, DataFrame Group)>
{
    private readonly DataFrame _df;
    private readonly LazyGroupBy _lazyGrouped;

    internal GroupByBuilder(DataFrame df, LazyGroupBy lazyGrouped)
    {
        _df = df;
        _lazyGrouped = lazyGrouped;
    }

    /// <summary>
    /// Filter groups with a predicate after aggregation.
    /// </summary>
    public GroupByBuilder Having(Expr predicate)
    {
        _lazyGrouped.Having(predicate);
        return this; 
    }

    /// <summary>
    /// Count the number of values in each group.
    /// </summary>
    public DataFrame Count()
        => Agg(Pl.All().Count());

    /// <summary>
    /// Aggregate all columns into lists. 
    /// </summary>
    public DataFrame All()
        => Agg(Pl.All()); 
    
    /// <summary>
    /// Aggregate the first values in the group.
    /// </summary>
    /// <param name="ignoreNulls">Ignore null values (default False). If set to True, the first non-null value for each aggregation is returned, 
    /// otherwise None is returned if no non-null value exists.</param>
    /// <returns></returns>
    public DataFrame First(bool ignoreNulls=false)
        => Agg(Pl.All().First(ignoreNulls)); 
    /// <summary>
    /// Aggregate the last values in the group.
    /// </summary>
    /// <param name="ignoreNulls">Ignore null values (default False). If set to True, the last non-null value for each aggregation is returned, 
    /// otherwise None is returned if no non-null value exists.</param>
    /// <returns></returns>
    public DataFrame Last(bool ignoreNulls=false)
        => Agg(Pl.All().Last(ignoreNulls)); 
    public DataFrame Head(int n = 10) => _lazyGrouped.Head(n).Collect();
    public DataFrame Tail(int n = 10) => _lazyGrouped.Tail(n).Collect();
    /// <summary>
    /// Return the number of rows in each group.
    /// </summary>
    /// <param name="name">Assign a name to the resulting column; if unset, defaults to “len”.</param>
    /// <returns></returns>
    public DataFrame Len(string name="len")
        => Agg(Pl.Len().Alias(name));
    /// <summary>
    /// Reduce the groups to the maximal value.
    /// </summary>
    /// <returns></returns>
    public DataFrame Max()
        => Agg(Pl.All().Max());
    /// <summary>
    /// Reduce the groups to the minimal value.
    /// </summary>
    /// <returns></returns>
    public DataFrame Min()
        => Agg(Pl.All().Min());
    /// <summary>
    /// Reduce the groups to the median value.
    /// </summary>
    /// <returns></returns>
    public DataFrame Median()
        => Agg(Pl.All().Median()); 
    /// <summary>
    /// Reduce the groups to the mean value.
    /// </summary>
    /// <returns></returns>
    public DataFrame Mean()
        => Agg(Pl.All().Mean()); 
    /// <summary>
    /// Count the unique values per group.
    /// </summary>
    /// <returns></returns>   
    public DataFrame NUnique()
        => Agg(Pl.All().NUnique());
    /// <summary>
    /// Reduce the groups to the sum.
    /// </summary>
    /// <returns></returns>  
    public DataFrame Sum()
        => Agg(Pl.All().Sum());  
    /// <summary>
    /// Compute the quantile per group.
    /// </summary>
    /// <param name="quantile">Quantile between 0.0 and 1.0.</param>
    /// <param name="interpolation">Interpolation method.</param>
    /// <returns></returns>          
    public DataFrame Quantile(double quantile,QuantileMethod interpolation = QuantileMethod.Linear)
        => Agg(Pl.All().Quantile(quantile,interpolation));   

    /// <summary>
    /// Aggregate with specified expressions.
    /// Under the hood, this routes through the Lazy engine to maximize performance and optimizations.
    /// </summary>
    /// <param name="aggs">Aggregation expressions</param>
    /// <returns>A new aggregated DataFrame</returns>
    public DataFrame Agg(params Expr[] aggs)
        => _lazyGrouped.Agg(aggs).Collect();

    private (DataFrame GroupsDf, string IndicesColName) BuildGroupsDataFrame()
    {
        string rowIdxCol = "__POLARS_GB_ROW_INDEX";
        string tempCol = "__POLARS_GB_GROUP_INDICES";

        var lazyWithIndex = _df.Lazy().WithRowIndex(rowIdxCol);
        var intoKeys = _lazyGrouped._keys.Select(k => (IntoExpr)k).ToArray();

        LazyGroupBy newLazyGrouped = _lazyGrouped._type switch
        {
            GroupByType.Dynamic => lazyWithIndex.GroupByDynamic(
                indexColumn: _lazyGrouped._indexColumn!,
                every: _lazyGrouped._every!,
                period: _lazyGrouped._period!,
                offset: _lazyGrouped._offset!,
                label: _lazyGrouped._label,
                includeBoundaries: _lazyGrouped._includeBoundaries,
                closedWindow: _lazyGrouped._closedWindow,
                startBy: _lazyGrouped._startBy,
                groupBy: intoKeys),

            GroupByType.Rolling => lazyWithIndex.Rolling(
                indexColumn: _lazyGrouped._indexColumn!,
                period: _lazyGrouped._period!,
                offset: _lazyGrouped._offset!,
                closedWindow: _lazyGrouped._closedWindow,
                groupBy: intoKeys),

            _ => lazyWithIndex.GroupBy(intoKeys, _lazyGrouped._maintainOrder)
        };

        var groupsDf = newLazyGrouped
            .Agg(Pl.Col(rowIdxCol).Implode().Alias(tempCol))
            .Collect();

        return (groupsDf, tempCol);
    }

    /// <summary>
    /// Allows iteration over the groups of the group by operation.(Strong Typed)
    /// </summary>
    public IEnumerable<(TKey Key, DataFrame Group)> GetGroups<TKey>() where TKey : new()
    {
        var (groupsDf, tempCol) = BuildGroupsDataFrame();
        
        using (groupsDf)
        using (var indicesCol = groupsDf[tempCol])
        using (var keysDf = groupsDf.Drop(tempCol))
        {
            var typedKeys = keysDf.Rows<TKey>().ToArray();

            for (int i = 0; i < typedKeys.Length; i++)
            {
                using var slicedList = indicesCol.Slice(i, 1);
                using var indices = slicedList.Explode(); 
                
                var groupDf = _df[indices]; 

                yield return (typedKeys[i], groupDf);
            }
        }
    }

    /// <summary>
    /// Allows iteration over the groups of the group by operation.
    /// </summary>
    public IEnumerator<(object[] Key, DataFrame Group)> GetEnumerator()
    {
        var (groupsDf, tempCol) = BuildGroupsDataFrame();

        using (groupsDf)
        using (var indicesCol = groupsDf[tempCol])
        using (var keysDf = groupsDf.Drop(tempCol))
        {
            for (int i = 0; i < keysDf.Height; i++)
            {
                object[] keyObjects = keysDf.Row(i)!;
                
                using var slicedList = indicesCol.Slice(i, 1);
                using var indices = slicedList.Explode(); 
                
                var groupDf = _df[indices];

                yield return (keyObjects, groupDf);
            }
        }
    }
    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
}